### ISSUE TITLE
[REQUEST] Categorical critic (C51)

### ISSUE THREAD

**Issue by alxlampe** [takuseno/d3rlpy#19]
Dear @takuseno,
I was trying build a categorical critic (c51) from the paper  [A Distributional Perspective on Reinforcement Learning](https://arxiv.org/pdf/1707.06887.pdf) into your d3rlpy library. 
But after inspecting the interfaces of your implemented Q-functions, I think it is not possible without much refactoring since C51 outputs a reward distribution with supports (atoms) instead of reward values or quantiles like all of your implemented Q-functions do.

For example, `SACImpl.compute_target()` expects that `targ_q_func.compute_target()` outputs reward values:
https://github.com/takuseno/d3rlpy/blob/523f89a52a1c2365a12043d6fbd9af3761ac288d/d3rlpy/algos/torch/sac_impl.py#L102-L106

Am I right or can you give me a hint, how to implement a categorical critic into your library?
Thank you very much.

**Comment by takuseno** [comment]
@alxlampe Hello, thank you for reporting the issue. Actually, C51 can be implemented very easily.

Before explaining how to make it, let me tell you why C51 is not available in d3rlpy. The reason is that the C51 is not actually a practical algorithm because it needs to tune the range of values, `V_max` and `V_min`. Tuning this is quite hard when trying to solve new environments. After Quantile Regression DQN, there is no need to do this, and the newer Q functions are actually working well without tuning their hyper parameters.

Then, let me give you some hints and I'll welcome your PR if you make it, of course!
You can just make `DiscreteCategoricalQFunction` and `ContinuousCategoricalQFunction` by refering the other Q functions.
https://github.com/takuseno/d3rlpy/blob/523f89a52a1c2365a12043d6fbd9af3761ac288d/d3rlpy/models/torch/q_functions.py#L193

And, you can make `compute_target` return `z_j` in below. And, you can finally make `compute_error` and its `q_tp1` will be the `z_j` from `compute_target`.
![image](https://user-images.githubusercontent.com/5235131/101013139-2d415900-35a7-11eb-997e-9175cd188b07.png)

**Comment by alxlampe** [comment]
Thanks for your reply @takuseno!

> Before explaining how to make it, let me tell you why C51 is not available in d3rlpy. The reason is that the C51 is not actually a practical algorithm because it needs to tune the range of values, `V_max` and `V_min`. Tuning this is quite hard when trying to solve new environments. After Quantile Regression DQN, there is no need to do this, and the newer Q functions are actually working well without tuning their hyper parameters.

That is what I read everywhere and why I thought, that using you library would improve my training, since I am currently using a self implemented C51 combined with Soft Actor Critic. 
But I do not get the performance with d3rlpy that I get with my implementation. I doesn't converge and my implementation gets better results after around 5000 train steps that I'll get using d3rlpy for e.g. 150000 steps. I have tried various learning rates, reward scaling and reward clipping but nothing worked out. 

I must say, that I am not using your library solely. I have integrated the main parts of your library into my framework because I have around 20 different environments for collecting training data and also 20 different environments for evaluation with some post processing etc. so that it would be a lot of effort to change everything to be compatible with your library. 
But I have tested everything with `Pendulum-v0` env to make sure that your library works in my framework.  Perhaps, this is not enough. 

My reward function is also very complex and I think difficult to approximate. I have 7 different reward terms where some of them are local constraints that can cause high costs if violated (I am working on a continuous control problem). 

My experience is (with my current implementation) that C51 is very stable, converging very fast and that it is not very sensitive to hyperparameters. And I experienced that I can use n_step returns with n=6  or higher which also improves convergence speed much.

So I am still wondering, why C51 is performing so much better than the advanced quantile regression methods, that you implemented. 
One thing could be, that by setting `v_min` and `v_max` the return is clipped which is good with regard to my spiky reward function. So I tried different things when using your library:
- clipping the reward -> no success
- implement clipping of the target used in the TD error calculation (q_tp1* gamma + r_tp1) to mimic the clipping of the categorical critic. I thought, that clipping the TD error is better than clipping the reward because one might have less loss of information. But actually also no success.

That was a bit off-topic but that is the reason why I want to implement it. If I will see poor performance with C51 in your library, the problem must be somewhere else. 


> Then, let me give you some hints and I'll welcome your PR if you make it, of course!
> You can just make `DiscreteCategoricalQFunction` and `ContinuousCategoricalQFunction` by refering the other Q functions.
> https://github.com/takuseno/d3rlpy/blob/523f89a52a1c2365a12043d6fbd9af3761ac288d/d3rlpy/models/torch/q_functions.py#L193

Thank you very much for your first hint. My C51 implementation looks quite different. It is based on the implementation from a book (Deep Reinforcement Learning Hands-On from Maxim Lapan).  But the algorithm you posted looks much simpler. 
I will send a PR that we can iterate on :)

**Comment by takuseno** [comment]
I see. That's interesting. Definitely, applying distributional Q functions to continuous control is not well investigated. And, it sounds like making sense that the clipped value range in C51 stabilizes the training in your scenario (I guess this is the result of good tuning).

> Thank you very much for your first hint. My C51 implementation looks quite different. It is based on the implementation from a book (Deep Reinforcement Learning Hands-On from Maxim Lapan). But the algorithm you posted looks much simpler.
I will send a PR that we can iterate on :)

Sounds great! Looking forward to that:)

**Comment by alxlampe** [comment]
I am still uncertain how to implement this in your library. 

> And, you can make `compute_target` return `z_j` in below. And, you can finally implement `compute_error` and its `q_tp1` will be the `z_j` from `compute_target`.

As far as I understand `z` is the support vector with its elements `z_j`, where `z` is defined with `v_min`, `v_max` and `n_atoms`:
```python
z =  torch.linspace(v_min, v_max, n_atoms)
```
So `z_j` values are constant.  
In the algorithm you depicted above, what we should get from our target network in `compute_target` is the probability `p(obs_tp1, a*)`  that is used in the algorithm to update `m_l` and `m_u`.

In my implementation, I have to incorporate the entropy in the projection and I see no other options to add it somewhere else.
To calculate the TD-error with entropy we use the following equation:
```
entropy  = - temp *  log(pi(s'))
Q(s,a) = r(s,a) + gamma * (Q(s',a') + entropy)

# results in:
Q(s,a) = r(s,a) + gamma * entropy + gamma * Q(s',a')
```
`r(s,a)` and `entropy` are both scalar values and `Q(s,a)` and `Q(s',a')` are both categorical distributions.  So we have to use `r(s,a) + gamma * entropy` in the projection.

Here are some snippets of my implementation of the projection with some explanations, we can discuss on.
The implementation is mainly from the book Reinforcement Learning Hands-on.
**Compute categorical distribution for next state**:
`next_action` is obtained from the target policy.
```python
next_state_dist = F.softmax(q_func(next_state, next_action, dim=1)
```
**Compute projection**:
```python
# projection with numpy is much faster than with torch
next_state_dist = next_state_dist.detach().cpu().numpy()
rewards = rewards.detach().cpu().numpy().squeeze()  # dim fix with squeeze, must be (batch_size,)
dones = dones.detach().cpu().numpy().astype(np.bool).squeeze()  # dim fix


if entropy is not None:
    # incorporate entropy in the projection. Reward values afterwards contain entropy values
    entropy = entropy.detach().cpu().numpy().squeeze()
    rewards = rewards + gamma * entropy * (1 - dones)

batch_size = len(rewards)
# initialize projected distribution
projected_dist = np.zeros([batch_size, self.n_atoms], dtype=np.float32)
# projection for-loop. Calculate projection for each atom
for atom in range(self.n_atoms):
    # compute location on 'x-axis' and find lower and upper support index
    tz_j = np.minimum(self.v_max,
                      np.maximum(self.v_min, rewards + (self.v_min + atom * self.delta) * gamma))
    b_j = (tz_j - self.v_min) / self.delta
    l = np.floor(b_j).astype(np.int64)
    u = np.ceil(b_j).astype(np.int64)
    
    # projection of the distribution (1. general case and 2. special case)
    # 1. general case where projected reward lands between to atoms
    ne_mask = u != l
    projected_dist[ne_mask, l[ne_mask]] += next_state_dist[ne_mask, atom] * (u - b_j)[ne_mask]
    projected_dist[ne_mask, u[ne_mask]] += next_state_dist[ne_mask, atom] * (b_j - l)[ne_mask]

    # 2. special case where projected reward lands on an atom and and general projection from above fails
    eq_mask = u == l
    projected_dist[eq_mask, l[eq_mask]] += next_state_dist[eq_mask, atom]
```
**Terminal states**: 
In terminal states, there is no `next_state_dist`. Landing in a terminal state, we get a terminal reward with probability=1 and no distribution. As a consequence, we are assigning a probability of 1 to the projected_dist if the projected reward lands on an atom or we spread the probability between two atoms:
```python
if dones.any():
    projected_dist[dones] = 0.0

   # find the value on the 'x-axis'  and the lower and upper atom
    tz_j = np.minimum(self.v_max, np.maximum(self.v_min, rewards[dones]))
    b_j = (tz_j - self.v_min) / self.delta

    l = np.floor(b_j).astype(np.int64)
    u = np.ceil(b_j).astype(np.int64)

      # 1. general case where projected reward lands between to atoms
    ne_mask = u != l
    ne_dones = dones.copy()
    ne_dones[dones] = ne_mask
    if ne_dones.any():
        projected_dist[ne_dones, l[ne_mask]] = (u - b_j)[ne_mask]
        projected_dist[ne_dones, u[ne_mask]] = (b_j - l)[ne_mask]

    # 2. special case where the projected rewards lands on an atom
    eq_mask = l == u
    eq_dones = dones.copy()
    eq_dones[dones] = eq_mask
    if eq_dones.any():
        projected_dist[eq_dones, l[eq_mask]] = 1.0
```

What do you think @takuseno? Am I still missing something?

**Comment by takuseno** [comment]
@alxlampe Oh, I see. I just missed that. As you initially said, it is not possible to implement C51 in d3rlpy for now. Then, I would say that d3rlpy possibly never supports C51. The current designs divides `algorithm loss` and `Q function loss` in a very sophisticated way, which enables us to use arbitrary Q functions on every algorithm. I don't think I can support C51 while maintaining this.

**Comment by alxlampe** [comment]
@takuseno I can understand. I also like your current design and wouldn't change it. If I have a good idea someday how C51 could be added, I will come back here :)

**Comment by takuseno** [comment]
Sorry for the convenience. But, sounds good:)

**Comment by takuseno** [comment]
I'll support C51 in the next release.

**Comment by GilgameshD** [comment]
I am also trying to use the C51 algorithm in this awesome package. Is it already implemented or stuck by obstacles?
